<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Crass: News</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <!-- Le styles -->
    <link href="bootstrap/css/bootstrap.css" rel="stylesheet">
    <style type="text/css">
      body {
        padding-top: 60px;
        padding-bottom: 40px;
      }
    small.crass {
      font-size: 75%;
    }
    </style>

    <!-- Le HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le fav and touch icons -->
    <link rel="shortcut icon" href="../assets/ico/favicon.ico">
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/ico/apple-touch-icon-144-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/ico/apple-touch-icon-114-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/ico/apple-touch-icon-72-precomposed.png">
    <link rel="apple-touch-icon-precomposed" href="../assets/ico/apple-touch-icon-57-precomposed.png">
  </head>

  <body>

    <div class="navbar navbar-fixed-top">
      <div class="navbar-inner">
        <div class="container">
          <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </a>
          <a class="brand" href="index.html">Crass</a>
          <div class="nav-collapse">
            <ul class="nav">
              <li><a href="assets/manual.pdf">Manual</a></li>
              <li><a href="news.html">News</a></li>
              <li><a href="http://ecogenomic.org">Ecogenomic</a></li>
            </ul>
          </div><!--/.nav-collapse -->
        </div>
      </div>
    </div>


      <div class="container-fluid">
      <div class="row-fluid">
        <div class="span3">
          <div class="well sidebar-nav">
            <ul class="nav nav-list">
			  <li class="nav-header">Latest Release</li>
              <li><a href="https://github.com/ctSkennerton/crass/archive/v0.3.5.tar.gz">Crass 0.3.5</a></li>
<li> <small class="crass">Skennerton, C. T., Imelfort, M. &amp; G. W. Tyson. (2013) "Crass: identification and reconstruction of CRISPR from unassembled metagenomic data" Nucleic Acids Research 41(10):e105; <a href="http://nar.oxfordjournals.org/content/41/10/e105.long">doi:10.1093/nar/gkt183</a> </small> </li>
              <li class="nav-header">Documentation</li>
              <li><a href="Tutorial.html">Tutorial</a></li>
              <li><a href="FAQ.html">Frequently Asked Questions</a></li>
              <li><a href="assets/manual.pdf">Manual</a></li>
              <li><a href="crass1.html">crass(1)</a></li>
              <li><a href="crass-assembler1.html">crass-assembler(1)</a></li>

              <li class="nav-header">Related Software</li>
              <li><a href="http://ctskennerton.github.com/crisprtools/">crisprtools</a></li>
              <li><a href="http://ctskennerton.github.com/libcrispr/">libcrispr</a></li>
              <li class="nav-header">Resources</li>
              <li><a href="assets/Crass-ISME14.pdf">ISME14 Poster</a></li>
            </ul>
          </div><!--/.well -->
        </div><!--/span-->
        <div class="span9">
          
          <div class="row-fluid">
            <div class="span9">
              <h3>Version 0.3.5 REleased</h3>
                <p>This is a bug fix release that resolves some problems caused in
                   version 0.3.4.  Please update to 0.3.5 to get the full benifits
                   that 0.3.4 was supposed to provide.
                </p>
              <h3>Version 0.3.4 Released</h3>
                <p>This is a bug fix release, but it may have important implications
                  for the interpretation of previous results!  I noticed an error on
                  one of my datasets which was causing crass to crash.  Upon tracking
                  it down I found that in certain edge cases a direct repeat would be 
                  aligned to another in such a way that would cause the alignment function to return 
                  an invalid value. 
                </p>

                <p>The actual cause was due to the aligner not checking
                to make sure that the aligned bases were not some internal motif
                that may be common to the DR group or including sequences that were
                close but not really from the DR.  This error was not caught at this stage but instead
                created flow on effects eventually resulting in a seemingly unrelated
                error in a different part of the code.
                </p>  
                <p>Aligning to an internal motif seemed to be the dominant alignments
                produced by repeats that were not of CRISPR origin and would get thrown 
                out anyway during later filtering steps in crass.  However
                I have noticed  that many groups have
                a few less spacers in them than they had before. I believe that this
                is a good thing, ie those 'spacers' were not real. However I have not
                performed extensive testing to determine if this is the case.  I 
                recommend all users to examine the results from both 0.3.4 (or greater)
                and versions 0.3.3 and before to see if there are spacers going missing.
                If there are manually check to see if those spacers are being removed
                improperly (ie crass shouldn't be throwing them out), in which case 
                open an issue in <a href="https://github.com/ctSkennerton/crass/issues"> github</a>
                or send me an email so I can work on the logic some more. 
                </p>
              <h3>Crass paper published</h3>
              <p>Our manuscript describing the Crass algorithm has been published in Nucleic Acids Research.  You can find the manuscript <a href="http://nar.oxfordjournals.org/content/early/2013/03/19/nar.gkt183.abstract">here</a></p>
              <h3>Version 0.3.0 Released</h3>
              <p>This version implements a number of changes, both 'under-the-hood' and visible to the user. Changes visible to the user are:</p>
              <ul>
                <li>All options now have a long and short form.  The new syntax is: </li>
                <table class="table">
                  <tr><th>long form</th><th>short form</th></tr>
                  <tr><td><code>--noRendering</code></td>  <td><code>-r</code></td></tr>
                  <tr><td><code>--removeHomopolymers</code></td>  <td><code>-H</code></td></tr>
                  <tr><td><code>--noDebugGraph</code></td>  <td><code>-e</code></td></tr>
                  <tr><td><code>--noScalling</code></td> <td><code>-z</code></td></tr>
                  <tr><td><code>--logToScreen</code></td><td><code>-g</code></td></tr>
                </table>
                <li>The assembly wrapper is now a separate executable.  You now must invoke it with <code>crass-assembler</code>
                  <p>Some other interface elements have also been changed in <code>crass-assembler</code>, the assembly wrapper of choice is now invoked as an option.  For example <code>crass-assembler --velvet</code> to run the velvet wrapper, or <code>crass-assembler --cap3</code> to run the cap3 wrapper.</p>
                </li>
                <li>The flow of Crass has changed.  In version 0.2.17 and below the basic layout of the search algorithm was: training, exhaustive search, cluster, assemble.  The new flow goes training, cluster, exhaustive search, assemble.  This may improve performance.</li>
                <li>The man pages have been updated (they were quite horribly inaccurate - sorry guys)</li>
              </ul>
              <p>There have been other changes 'under-the-hood' to the search algoriths and to the alignment of DR types for determining the correct boundaries for a direct repeat.  This means that you may find Crass finding more reads or even new DR types for the same datasets!!  It may be worth rerunning Crass on any datasets that you've processed to see the differences</p>
              <h3>Crass at ISME14</h3>
              <p>Crass was presented as a poster at the 14th international symposium on micrbial ecology in beautiful Copenhagen, Denmark.  See the pdf of the poster <a href="assets/Crass-ISME14.pdf">here</a> </p>
            </div><!--/span-->

          </div><!--/row-->
        </div><!--/span-->
      </div><!--/row-->
      <hr>

      <footer>
        <p>&copy; Connor Skennerton &amp; Michael Imelfort 2011 - 2013</p>
      </footer>

    </div> <!-- /container -->

    <!-- Le javascript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="bootstrap/js/jquery.js"></script>
    <script src="bootstrap/js/bootstrap-transition.js"></script>
    <script src="bootstrap/js/bootstrap-alert.js"></script>
    <script src="bootstrap/js/bootstrap-modal.js"></script>
    <script src="bootstrap/js/bootstrap-dropdown.js"></script>
    <script src="bootstrap/js/bootstrap-scrollspy.js"></script>
    <script src="bootstrap/js/bootstrap-tab.js"></script>
    <script src="bootstrap/js/bootstrap-tooltip.js"></script>
    <script src="bootstrap/js/bootstrap-popover.js"></script>
    <script src="bootstrap/js/bootstrap-button.js"></script>
    <script src="bootstrap/js/bootstrap-collapse.js"></script>
    <script src="bootstrap/js/bootstrap-carousel.js"></script>
    <script src="bootstrap/js/bootstrap-typeahead.js"></script>
